### DISCUSSION TITLE
How to update SystemWebView with new name ?

### DISCUSSION THREAD

**Discussion by frank-github** [bromite/bromite#1809] [Q&A]
I'm running my system for long time with bromite systemwebview on LineageOS.
It was installed with nanodroid (seem not be maintained anymore?) 
I update bromite systemwebview over F-Droid Repro on regular
bromite systemwebview is the only webview provider on system.

Now package name changed and f-droid install this package, but it's not recognized as systemwide webview provider.
System still use the old one (v96.xx)
I tried to install apk manually, but still v97 was not recognized as webview provider
I tried to replace apk with root file manager als written [here](https://github.com/bromite/bromite/wiki/Installing-SystemWebView), but as result system don't have a webview provider anymore
After restoring the old version of bromite systemwebview apk system recognized this as provider again.

How can I upgrade to this new version as a systemwide webview provider?

**Comment by joeo10** [comment]
I have a similar issue.

The current installation and wiki doesn't work with Android 11+ devices (as I have a A11 device) as it's now located on product/app/webview/webview.apk instead of system/app/webview/webview.apk

I tried putting the WebView onto system but it fails every time. Tried putting into product and it doesn't show up after rebooting. It used to work every time on the old package name on product/app/webview/webview.apk bit once the package name changes it doesn't work anymore. I hope there's a fix or workaround for this.

**Reply by csagan5** [reply]
Which ROM are you using?

**Reply by joeo10** [reply]
Bliss ROM and PixelDust, both Android 11 on Pixel 2 XL and 4 XL.

**Reply by Krax-001** [reply]
I also have this issue. After version 96 of webview it doesn't show up in WebView Implementation option under developer's options. v96 works just fine but latest versions breaks the app. I am currently on android 12 and been using google webview.

**Reply by csagan5** [reply]
> The current installation and wiki doesn't work with Android 11+ devices (as I have a A11 device) as it's now located on product/app/webview/webview.apk instead of system/app/webview/webview.apk

Feel free to update the wiki if you have reliable information on this.

> Bliss ROM and PixelDust, both Android 11 on Pixel 2 XL and 4 XL.

Did you follow the wiki instructions at https://github.com/bromite/bromite/wiki/Installing-SystemWebView#verifying-allowed-webview-packages? What does the `config_webview_packages.xml` contain?

These scenarios are possible:
* ROM is using an unsecure webview configuration (no signature)
* ROM was using the Bromite signature for it
* you installed it using some other different hack

In any case the ROM maintainer should be able to tell you.

**Comment by frank-github** [comment]
I'm using Android 9, LineageOS 16
and location same as @joeo10

**Comment by sergiok707** [comment]
I have the same problem, I can't select V97, V98 anymore
My phone no longer recognizes it as a WebView service provider

**Comment by sergiok707** [comment]
Will all new versions have this package name now? And I can no longer use the new version?

**Reply by frank-github** [reply]
it would be nice if the developer could give hints on how to fix the problem now.
He will surely have had his good reasons to rename the package. But it is very unpleasant that so far satisfied users can apparently no longer use his product.

Using outdated WebView versions is not necessarily advisable. One reason why I was so enthusiastic about Bromite was that it allowed me to keep WebView up to date, even though my ROM has not been maintained for a long time.

**Reply by csagan5** [reply]
You need to modify your ROM in order to use the new SystemWebView package name. If your ROM allows the old package name without signatures it is **unsecure**.

Everything is explained in the wiki.

**Reply by frank-github** [reply]
@csagan5 
as I wrote above, I tried it like explanation in wiki
> I tried to replace apk with root file manager als written [here](https://github.com/bromite/bromite/wiki/Installing-SystemWebView), but as result system don't have a webview provider anymore

it look's like I'm not the only one with problems. 
I would like to fix the problem, but I lack the information on how to fix it.
The description in the wiki does not work for me. Bromite WebView worked fine for a long time (incl. updates!).
Only with the new name it doesn't work anymore.

I can run adb commands, but I am neither Android developer nor ROM maintainer.

One thought, I don't know the effort, but would it possibly be possible to release the code as two separate packages? Once forward-looking with the new name and for old ROMs (like mine) still with the old name?
I think this would help some people :-)

**Reply by csagan5** [reply]
You need to modify a file inside the ROM in order to fix this, it is mentioned in the wiki. Sorry, two versions cannot be kept.

**Reply by frank-github** [reply]
Sorry if I disagree.
But the description in the wiki does not work.
If I overwrite the previous WebView.apk with the new version (exactly as described in the wiki) with root privileges or via recovery, then the system does not recognize the WebView implementation afterwards.
Some important step seems not to be listed in the wiki.
https://github.com/bromite/bromite/wiki/Installing-SystemWebView#installation-with-root-file-manager-or-twrp

before renaming of bromite webview, bromite webview was recognized and worked smooth and fine

**Reply by csagan5** [reply]
You have skipped one step: https://github.com/bromite/bromite/wiki/Installing-SystemWebView#verifying-allowed-webview-packages

**Reply by frank-github** [reply]
but I have made ;-)

bromite is not listed there.
Yes, now you can answer that the wiki says that then bromite cannot be used as WebView provider.

But I still use the now obsolete version of bromite successfully !
Unfortunately I can't say how nanodroid managed this, but it works smoothly (incl. earlier with the updates)
And I would now like to continue using it :-)

Is there no adb command how to install a new WebView provider and make it known to the system?
unfortunately I have had no luck so far with my internet research on this. But I'm probably searching with the wrong keywords.

The nanodroid installation script actually looks very clear, but then my Android knowledge is not enough :-(
https://downloads.nanolx.org/NanoDroid/Stable/NanoDroid-BromiteWebView-23.1.2.20210117.zip

**Reply by csagan5** [reply]
> but I have made ;-)
> 
> bromite is not listed there. Yes, now you can answer that the wiki says that then bromite cannot be used as WebView provider.

If you have read that the SystemWebView needs to be in that file, then why do you pretend you have not?

> unfortunately I have had no luck so far with my internet research on this. But I'm probably searching with the wrong keywords.
> 
> The nanodroid installation script actually looks very clear, but then my Android knowledge is not enough :-( https://downloads.nanolx.org/NanoDroid/Stable/NanoDroid-BromiteWebView-23.1.2.20210117.zip

You are on your own; the SystemWebView is provided as-is and it is for advanced users that can figure out how to install it. No support is provided here for that; if the instructions on the wiki are not sufficient you can search online how to install any SystemWebView, the procedure is not any different for the Bromite SystemWebView.

**Reply by sergiok707** [reply]
I also still can't seem to install the newer versions of Bromite as a WebView(

**Reply by sergiok707** [reply]
Buddy, I will be very grateful if you send 1 link, which discusses how to install any SystemWebView, I did not find it in Google

**Reply by frank-github** [reply]
@sergiok707 @Krax-001 @joeo10 
I found a solution to install the new version on my system --> Magisk WebViewManager did the job :1st_place_medal:

**Comment by sergiok707** [comment]
Dear developer, maybe you could return the old package name?
Thanks

**Comment by LeOS-GSI** [comment]
Yes, I do have the same issue on my LeOS-GSI builds. No way to include it since the package name is changed. I don*t know why it was done, but PLEASE CHANGE IT BACK

**Comment by lss4** [comment]
This name change brings the following advantages:
- Bromite SystemWebView no longer requires to be installed as system app anymore.
- Stock webview and Bromite can now coexist. In case Bromite WebView has issues, you can safely fall back to stock WebView.

The core issue of this name change, however, is how to patch `/system/framework/framework-res.apk` to add the necessary entry into `res/xml/config_webview_packages.xml` in a systemless and ROM-independent manner.

You can always patch the file of your current ROM yourself and put it into an installed and enabled Magisk Module to let it systemlessly replace the file, but the patch would still not be portable: You most likely need to remove the patch when you flash/update your ROM or you risk bricking.

EDIT: I'm currently looking at Androidacy's WebView Manager which frank-github mentioned in a previous post. ~It seems to be doing the right thing using an overlay.~

EDIT 2: Nope... the overlay generated by WebView Manager doesn't appear to insert the signature block. I think this will only work if installed as system app. To install Bromite WebView as pure user app the signature block is required.

**Reply by csagan5** [reply]
I will mention these advantages in the release news; as for the installation methods: it is my understanding that the new package name will not be less secure than the old one, possibly more. ROMs either need to embed the signature block or manual modifications are needed, but the same was true for the previous package name as well.

**Reply by lss4** [reply]
> it is my understanding that the new package name will not be less secure than the old one, possibly more.

Indeed. The new package name avoids most of the force closes (or even bootloops) caused by improper replacement of the stock WebView, as replacing system components is always risky.

> ROMs either need to embed the signature block or manual modifications are needed, but the same was true for the previous package name as well.

For the previous package name, modifications are needed on stock ROMs that do not already have the entry, since they ship with Google WebView or Chrome Stable by default. Custom ROMs (as well as GSIs) generally do have the stock WebView entry and do not need modifications.

For current package name, unfortunately, modifications are always required unless some ROM developers agreed to permit trustworthy third-party WebView providers by including their package names and official signatures, so users can download, install and use these providers as long as they're verified.

WebView Manager currently makes the changes to the XML via overlay, but it's not inserting the corresponding signatures and is using the `availableByDefault=true` flag instead. This works as it installs the same way as some others do: first systemlessly installs the apk as system app then install the same apk as user app after initial startup. However, this is less secure, as in this way, if the package initially installed has been tampered with, it would still be permitted.

EDIT: Some clarifications. It appears since Android 10 and onwards all defined WebView providers have the `availableByDefault=true` flag regardless, so the flag may not be as relevant as it was on Android 9 and before. However, a signature block is still needed if you want to use it while installing it as user app.

I just managed to create a functional overlay using the tools included in the WebView Manager, with Bromite's package name and signature added into the XML I extracted from my current ROM. After installing the overlay, I'm able to select the new Bromite WebView right after installing it as a user app, with no more actions needed.

**Reply by csagan5** [reply]
> Custom ROMs (as well as GSIs) generally do have the stock WebView entry and do not need modifications.

Is that because they can accept any `com.android.webview` package and not verify its signature?



> WebView Manager currently makes the changes to the XML via overlay, but it's not inserting the corresponding signatures and is using the `availableByDefault=true` flag instead. This works as it installs the same way as some others do: first systemlessly installs the apk as system app then install the same apk as user app after initial startup. However, this is less secure, as in this way, if the package initially installed has been tampered with, it would still be permitted.

Without a pinned signature and a pre-installed package it will allow any `com.android.webview` with any signature?

> I just managed to create a functional overlay using the tools included in the WebView Manager, with Bromite's package name and signature added into the XML I extracted from my current ROM. After installing the overlay, I'm able to select the new Bromite WebView right after installing it as a user app, with no more actions needed.

Can you suggest some improvements for the [wiki page](https://github.com/bromite/bromite/wiki/Installing-SystemWebView)? I cannot give edit access to individuals here on GitHub but you can clone the pages from `git@github.com:bromite/bromite.wiki.git` and propose the changes.

**Reply by lss4** [reply]
> 
> Is that because they can accept any `com.android.webview` package and not verify its signature?
> 
> Without a pinned signature and a pre-installed package it will allow any `com.android.webview` with any signature?
> 

`com.android.webview` is specified without a `signature` block. This allows any **system** package with that name (Stock/AOSP, Chromium, Ungoogled) to be usable as a WebView provider. **Packages that are purely user-installed will not be permitted this way, and subsequent package upgrades must have the matching signature of the initial one. For example, if you installed Ungoogled to system, then you can only upgrade with another Ungoogled package afterwards.**

A `signature` block is **required** in order to allow a purely user-installed WebView package to be usable as a provider. In this case, the user-installed WebView package will be usable **if it has the right name and right signature**.

> Can you suggest some improvements for the [wiki page](https://github.com/bromite/bromite/wiki/Installing-SystemWebView)? I cannot give edit access to individuals here on GitHub but you can clone the pages from `git@github.com:bromite/bromite.wiki.git` and propose the changes.

I'm not sure how to explain everything in a beginner friendly manner, as at present I'm yet to find a readily available tool that could automate the procedure in a correct and foolproof way.

**Should seriously warn that if `res/xml/config_webview_packages.xml` fails validation the system will BOOTLOOP. Care must be taken when editing that file, and you'll need to have a means to revert the changes in case anything wrong happens.**

For now you can use Androidacy WebView Manager to do the initial installation, but be noted that it doesn't add the signature block for the WebView provider you chose, and that it downloads the WebView package through its own web API, not directly from here.

I used the data files and tools inside Androidacy WebView Manager to create the overlay, but the procedure was kind of dirty and the overlay seems to have some minor side effects which affected the functionality of the module I used to carry the overlay. At least it proved "patching" the `framework-res.apk` using an overlay can work for this purpose (to make installed Bromite WebView usable).

**Reply by csagan5** [reply]
> > Is that because they can accept any `com.android.webview` package and not verify its signature?
> > Without a pinned signature and a pre-installed package it will allow any `com.android.webview` with any signature?
> 
> `com.android.webview` is specified without a `signature` block. This allows any **system** package with that name (Stock/AOSP, Chromium, Ungoogled) to be usable as a WebView provider. **Packages that are purely user-installed will not be permitted this way, and subsequent package upgrades must have the matching signature of the initial one. For example, if you installed Ungoogled to system, then you can only upgrade with another Ungoogled package afterwards.**

Ok, thanks - this matches how I understood it works.

> > Can you suggest some improvements for the [wiki page](https://github.com/bromite/bromite/wiki/Installing-SystemWebView)? I cannot give edit access to individuals here on GitHub but you can clone the pages from `git@github.com:bromite/bromite.wiki.git` and propose the changes.
> 
> I'm not sure how to explain everything in a beginner friendly manner, as at present I'm yet to find a readily available tool that could automate the procedure in a correct and foolproof way.
> 
> **Should seriously warn that if `res/xml/config_webview_packages.xml` fails validation the system will BOOTLOOP. Care must be taken when editing that file, and you'll need to have a means to revert the changes in case anything wrong happens.**

I have made some edits to highlight this.

> For now you can use Androidacy WebView Manager to do the initial installation, but be noted that it doesn't add the signature block for the WebView provider you chose, and that it downloads the WebView package through its own web API, not directly from here.
> 
> I used the data files and tools inside Androidacy WebView Manager to create the overlay, but the procedure was kind of dirty and the overlay seems to have some minor side effects which affected the functionality of the module I used to carry the overlay. At least it proved "patching" the `framework-res.apk` using an overlay can work for this purpose (to make installed Bromite WebView usable).

I cannot endorse any unofficial installation method because some of them use compiled binaries instead of building everything from source, but if you would like to add more information about the overlay approach I will be glad to add it to the wiki.

**Comment by frank-github** [comment]
> For now you can use Androidacy WebView Manager to do the initial installation, but be noted that it doesn't add the signature block for the WebView provider you chose, and that it downloads the WebView package through its own web API, not here.
I did use Androidacy WebView Manager too
Later updates could installed with apk downloaded here without any 
problems :-)

**Reply by lss4** [reply]
> I did use Androidacy WebView Manager too
> Later updates could installed with apk downloaded here without any
> problems :-) 

It installs the WebView as a system package systemlessly. This way doesn't require signature block, but given you can install updates from here on top of the one installed from WebView Manager it means the package provided by their API is authentic.

**Comment by lss4** [comment]
Great news for latest phh-AOSP GSI users:

As of phh-AOSP 12.1 v411, Bromite SystemWebview is usable out-of-box! After installing the Bromite SystemWebView APK on that GSI I can immediately choose it from the Developer Options menu and start using it. No other actions needed.

If you're currently on an Android 12 GSI dated prior to this, it may be possible to put that overlay package from phh v411 (`/system/product/overlay/treble-overlay-highpriomisc/treble-overlay-highpriomisc.apk`) into your GSI to get it working, though from the discussion it may not work on Android 11.

See [this](https://github.com/phhusson/treble_experimentations/issues/2239) and [this](https://github.com/phhusson/treble_experimentations/discussions/2236#discussioncomment-2408062) for details.

**Comment by ghost** [comment]
Hi. I recently installed the Bromite SystemWebView and I have no such a problem. Could you run these commands and send a screenshot after that? (If you want to do it on Windows, search for the solution on the web)


sudo apt install adb aapt -y

adb pull /system/framework/framework-res.apk

aapt d xmltree framework-res.apk res/xml/config_webview_packages.xml

**Reply by ghost** [reply]
Your problem should be solved If you can succesfully  edit the config_webview_packages.xml to include the text(open link to see it)

https://mega.nz/file/wmBBzDia#dljkRXhOh1autY5kj-TTVmvhhQv2qwrAUsZdrhyXgeg

https://mega.nz/file/ErRWTQ4L#1dKhVIIxieOP24gq8vTstgztOfYsCizD7xAeYJVluSg

If you have twrp you can even do that without root. 
Personally I decompiled the framework-res.apk edited the file, built the new apk afterwards and replaced the config_webview_packages.xml from the original apk with the file from the new apk.
Should I write a guide, or you can figure it out how to do that?

**Reply by ghost** [reply]
Maybe with root would work those apps, or in the future if they will get new features/fixes, but on my device currently those apps can't do the thing.

If you want to do it without root I can only suggest apktool on linux(not the apt version).
I done it with that.

**Reply by anyone00** [reply]
I think it would make more sense to put the instructions on the bromite wiki or put it in some kind of blog (or a gist/github repo if you don't have a blog).

**Reply by ghost** [reply]
First I would need to write the instructions and make it easy to understand, if there is need for it I will do it to help you guys. I do not speak the language too well, but I think I could make it simple enough soo everyone could follow it who has a bit of experience useing linux.
Or should I make a video as well?
I think I will probably do it this week (at least the instructions).

**Reply by anyone00** [reply]
It surely will be helpful!

About the language I will try to help correct issues if I find them.

Adding instructions for linux for now should be fine, I think if someone wants to improve it they can writes a windows guide too!

Do it the way you like it! Video, script, text ...etc
take your time :+1:

**Reply by ghost** [reply]
If you want the modified apk please send me a new one in an hour. The one which I received probably would not work, it has an invalid chunk type.
You have adb on windows use
adb pull /system/framework/framework-res.apk
Compress it to a zip and sent it that way.
If I do not get a correct apk before I go to sleep you will need to do it yourself following the instructions which will be available in the coming days.

**Reply by csagan5** [reply]
> I think I will probably do it this week (at least the instructions).

Let me know if I can change/add something to the [SystemWebView wiki page](https://github.com/bromite/bromite/wiki/Installing-SystemWebView)

**Reply by ghost** [reply]
After I made the guide, I will make it available for everyone. You can include it after that.

**Reply by LeOS-GSI** [reply]
it seems to be a device specific issue. On all my test devices and no other users in my LeOS-Support group does have this issue. The version number is always showing in dev. options
![webview](https://user-images.githubusercontent.com/2726574/164909432-36090664-7677-4f89-92f7-a0f519e07fd8.png)
g

**Comment by ghost** [comment]
@BiGroBB911 I still got an error with your framework-res.apk. It turns out that the tool I use is not compatible yet with the Android API level 32 (Android 12L). Currently, my method would only work for you if you would downgrade your ROM to the 31 API level (Android 12).
Android 12L users without root need to remain with the overlay method for a while which may not be perfect.

**Reply by ghost** [reply]
There is still one thing that you could try altho the chance of fixing your issue is low. Download and unzip the file https://file.io/o1i4KXIKgSOy
In TWRP on the first column, there is a section called Mount, in that system should be selected. Go back and to the Advanced section there is a File Manager with the help of that move the folder called bromite to /system/priv-app
(Your phone storage is called sdcard if you put the folder on the external sd card that is called external_sd)(In case it does not help you can later delete the folder with the same method)
In case this somehow helped and the version 100.0.4896.135 will appear in your settings please let us know.

**Reply by ghost** [reply]
Obviously, If your problem is already solved, there is no reason to do it. (the idea was if you install it as a system app maybe your device will see it because the overlay method already worked with other versions)

**Reply by ghost** [reply]
Didn't you wrote that you rooted the phone and your problem is solved? If you are talking about my last idea I wrote the chance of fixing the issue is low.
 On non rooted phones the only method which should granted work (unless there is a device or ROM specific problem) is modifying the framework-res.apk with your pc plus you need to have a custom recovery on your phone (ex: TWRP). The instructions for that method is not yet avaible and currently devices higher than android 6 but lower than android 12L are supprted. Sorry, I can't help you if you have an android 12L device. Maybe a few months later it will be possible to use my method on 12L devices if apktool gets an update or if someone could find a better tool.

**Reply by ghost** [reply]
I have no idea what did you done but I said my method is not available yet, the instructions are not done, and even if it would be currently it does not work on 12L devices.

If you are talking about the other advice (the idea that may do something on your device)
If you copy a folder into /system/priv-app and it does not work it can be simply removed

You wrote you used DisableForceEncryption.zip, you had twrp but you had enceyption on?  Why do you have twrp if you don't decrypt your device?
Does not the twrp tutorial included the process of decrypting your device?

**Reply by ghost** [reply]
I think I never had such a problem, or I just forgot, but after a quick search, I see many people faced that issue. On youtube there are short tutorials how to fix it, to make your twrp work properly (probably you have already done that). I am sorry that your OS is gone but I had no way to know that your twrp will not function properly.

**Comment by LeOS-GSI** [comment]
OK, guys, please stop wasting your time !!!!!!!! This is a self created issue !!!!! 
The user is playing with the 'dp' settings in developer option, that's why the number isn't shown. The easiest way to show the number is, using the device in landscape, because the 'dp' settings are wrong, the screen can't show the long number.
![potrait](https://user-images.githubusercontent.com/2726574/164913041-495391ad-94bf-46e6-8e18-3d8ef93cd032.png)

![landscape](https://user-images.githubusercontent.com/2726574/164913039-ec775b46-0a18-45d8-a64b-f61b20411256.png)

**Reply by ghost** [reply]
Thank you. I assumed he has a more serious problem. I assumed the webview does not work for him, I did not think he would make this big deal just about simply not seeing the version number.

**Reply by LeOS-GSI** [reply]
![IMG_20220423_191048_008](https://user-images.githubusercontent.com/2726574/164916445-5544fed0-491f-449d-8c1f-82919f51ea8d.jpg)
dumpsys is showing that his webview version is updated. So all fine. I know that, because he has posted the dumpsys result on a telegram group.

**Reply by LeOS-GSI** [reply]
I hope the creator of this issue will close it, because it is solved as he has posted in.Telegram group

**Reply by ghost** [reply]
I was able to resolve my issue, it was a dp setting issue.
Thank you very much for your kind help 😃👍
And sorry again for the inconvenience
![Screenshot_20220423-192604](https://user-images.githubusercontent.com/92749329/164944355-fa436f77-4a58-4691-bd37-7b0edbd2ce4d.png)

**Reply by ghost** [reply]
This discussion is not only related to his issue. 
Some other users with different ROMs are unable to install the Bromite SystemWebView, because in their framework-res.apk the config_webview_packages.xml does not contain the packagename="org.bromite.webview" and they don't know how to edit it, or use the overlay method. 
I managed to edit the framework-res.apk on my devices without the phones being rooted (currently works on devices with lower API level than 32). I was working on writing the instructions to solve their problem but I realized it would not be a user-friendly experience and would leave too much room for user error. So today I started to put the commands into scripts. I am fairly close to being done with it. If I will have enough free time in the next week I will even try to make a simple GUI for the scripts. In case the process fails on some devices, there will be a script which will restore the original framework-res.apk
The GUI or at least the scripts can be expected to be available in the next week.

**Comment by Lky777** [comment]
just like other guys said, u need edit _framework-res.apk/res/xml/config_webview_packages.xml_at first.

![IMG_20220515_212700](https://user-images.githubusercontent.com/73606056/168475270-ae45a76c-0e1d-4f52-900c-edf26885e4d1.jpg)

Tested OK on MIUI 12.5/Android 10.

![IMG_20220515_212627](https://user-images.githubusercontent.com/73606056/168475289-e42b4c4a-98b7-40a7-b678-6ff4598d48f6.jpg)

**Reply by LeOS-GSI** [reply]
I thought the issue/discussion is closed, because all is working now
